frontend/openchat-client/src/utils/cachePrimer.ts

```diff
@@ -0,0 +1,149 @@
+import type {
+    ChatEvent,
+    ChatSummary,
+    EventsResponse,
+    EventWrapper,
+    IndexRange
+} from "openchat-shared";
+import { compareChats, missingUserIds, userIdsFromEvents } from "openchat-shared";
+import { toRecord } from "./list";
+import { Poller } from "./poller";
+import type { OpenChatAgentWorker } from "../agentWorker";
+import { selectedChatId } from "../stores/chat";
+import { boolFromLS } from "../stores/localStorageSetting";
+import { messagesRead } from "../stores/markRead";
+import { userStore } from "../stores/user";
+import { get } from "svelte/store";
+
+export class CachePrimer {
+    private pending: Record<string, ChatSummary> = {};
+    private runner: Poller | undefined = undefined;
+    private selectedChatId: string | undefined = undefined;
+
+    constructor(private api: OpenChatAgentWorker) {
+        selectedChatId.subscribe((chatId) => {
+            if (chatId !== undefined) {
+                delete this.pending[chatId];
+            }
+            this.selectedChatId = chatId;
+        });
+        debug("initialized");
+    }
+
+    processChatUpdates(previous: ChatSummary[], next: ChatSummary[]): void {
+        const record = toRecord(previous, (c) => c.chatId);
+
+        const updated = next.filter((c) =>
+            c.chatId !== this.selectedChatId &&
+            !c.archived &&
+            hasBeenUpdated(record[c.chatId], c)
+        );
+
+        this.enqueue(updated);
+    }
+
+    processChatMarkedAsRead(chat: ChatSummary): void {
+        if (chat.chatId !== this.selectedChatId && !chat.archived) {
+            this.enqueue([chat]);
+        }
+    }
+
+    async processNext(): Promise<void> {
+        try {
+            const chat = Object.values(this.pending).sort(compareChats)[0];
+            if (chat === undefined) {
+                debug("queue empty");
```

Comment:
Do we still need all the debug statements?

Reply:
I guess I can take them out once I've seen it working for a while on webtest.